### PR TITLE
[DTensor] Support in gradient placement for local_map()

### PR DESCRIPTION
**Pull request by zpcore**
Support `in_grad_placements` argument in torch.distributed.tensor.experimental.local_map().  The argument helps enforce placement of gradient of the input Dtensor.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

### CODE REVIEW

File: test/distributed/tensor/experimental/test_local_map.py [hunk 1]
```diff
@@ -318,6 +324,90 @@ def test_local_map_redistribute(self):
         with self.assertRaisesRegex(ValueError, "set redistribute_inputs=True"):
             Y_dt = local_mm_allreduce_forward(device_mesh, X_dt, W_dt)
 
+    @with_comms
+    def test_local_map_with_grad_placement_modifiable(self):
```
Comment (XilunWu) [inline test/distributed/tensor/experimental/test_local_map.py]: IIUC this test solely checks that `in_grad_placements` can be correctly assigned to `.grad`. I think it can be merged into "test_local_map_with_grad_placement_correctness" i.e. we don't need a separate test to check this.

File: test/distributed/tensor/experimental/test_local_map.py [hunk 2]
```diff
@@ -318,6 +324,90 @@ def test_local_map_redistribute(self):
...
+
+        X_dt = distribute_tensor(X, device_mesh, row_wise)
+        W_dt = distribute_tensor(W, device_mesh, col_wise)
+
+        in_grad_placements = (row_wise, col_wise)
+        local_mm_forward = local_map(
+            mm_forward,
+            out_placements=[Partial()],
+            in_placements=(col_wise, row_wise),
+            in_grad_placements=in_grad_placements,
```
Comment (XilunWu) [inline test/distributed/tensor/experimental/test_local_map.py]: One problem of this test is, the test still passes even if we remove this `in_grad_placements`. Let's come up with another test case.

File: torch/distributed/tensor/experimental/_func_map.py [hunk 3]
```diff
@@ -19,14 +19,15 @@
 __all__ = ["local_map"]
 
 PlacementType = Optional[Sequence[Placement]]
-InputPlacements = Optional[tuple[PlacementType, ...]]
-OutputPlacements = Union[PlacementType, tuple[PlacementType, ...]]
+InputPlacements = Optional[Sequence[PlacementType]]
+OutputPlacements = Union[PlacementType, Sequence[PlacementType]]
```
Comment (XilunWu) [inline torch/distributed/tensor/experimental/_func_map.py]: Let's not make this change. We want to keep the style as:
```
(
    [Placement0, Placement1, ...],   # input 0
    [Placement0, Placement1, ...],   # input 1
    [Placement0, Placement1, ...],   # input 2
    ...
)
```

File: torch/distributed/tensor/experimental/_func_map.py [hunk 4]
```diff
@@ -66,6 +67,13 @@ def local_map(
             will be skipped and the argument will be directly passed to ``func``.
             If ``in_placements`` is ``None``, no placements examination will be performed.
             Default: None
+        in_grad_placements (Tuple[`PlacementType`, ...], optional):
+            the required placements of the :class:`DTensor` s gradient during back backpropagation.
+            It only works when input tensor to ``func`` is of type :class:`DTensor`. Input
+            :class:`DTensor` to ``func`` will be converted to tensor first with gradient
+            ``in_grad_placements`` enforced on the original :class:`DTensor`. Afterwards,
+            tensor will be converted back to :class:`DTensor` using original placement or
+            ``in_placements`` if specified.
```
Comment (XilunWu) [inline torch/distributed/tensor/experimental/_func_map.py]: This definition needs change, let's discuss offline.

**Review by XilunWu — COMMENTED** [review]
This PR is a good first step, but we need to further clarify some concepts and modify some part of the code. Let's discuss offline.

File: test/distributed/tensor/experimental/test_local_map.py (hunk 2 above)
Comment (zpcore) [reply test/distributed/tensor/experimental/test_local_map.py]: You are right, the test still passes because the `in_grad_placements` here is the same as original `X_dt` and `W_dt` placement. If we update  `in_grad_placements = (row_wise, col_wise)`, the test will fail as expected. I need to come up with a better test.

File: torch/distributed/tensor/experimental/_func_map.py (hunk 3 above)
Comment (zpcore) [reply torch/distributed/tensor/experimental/_func_map.py]: I see, updated the annotation.

File: test/distributed/tensor/experimental/test_local_map.py (hunk 1 above)
Comment (zpcore) [reply test/distributed/tensor/experimental/test_local_map.py]: Tried to showcase that `in_grad_placements` can work with `in_placements` here originally. I have merged the test with other test function.

**Comment by zpcore** [comment]
Hi @wanchaol , @Chillee , we are currently working on supporting the in_grad_placement for the `local_map()` and are wondering what is the use case here so that we can figure out if this is the correct design. Can you help provide context on why in_grad_placement is wanted? Thanks

**Comment by tianyu-l** [comment]
> what is the use case here so that we can figure out if this is the correct design. Can you help provide context on why in_grad_placement is wanted?

I feel this shouldn't be different from other cases when we need to call `to_local` with `grad_placements`. A typical example would in Data Parallel, different ranks could have `Replicate` parameters, but are seeing different plain tensor data thus having `Partial` gradients.

https://github.com/pytorch/pytorch/blob/main/test/distributed/tensor/test_dtensor.py#L350

File: torch/distributed/tensor/experimental/_func_map.py [hunk 5]
```diff
@@ -127,6 +135,42 @@ def local_map(
     .. note:: This API is currently experimental and subject to change
     """
 
+    def maybe_redistribute(
+        input_tensor: Union[DTensor, torch.Tensor],
+        device_mesh: DeviceMesh,
+        spec: PlacementType,
+        redistribute_inputs: bool = True,
```
Comment (wanchaol) [inline torch/distributed/tensor/experimental/_func_map.py]: The default `redistribute_inputs` value is False in `local_map`, but in this function it becomes true? It does not seems to aligned

File: torch/distributed/tensor/experimental/_func_map.py [hunk 6]
```diff
@@ -155,27 +199,32 @@ def wrapped(device_mesh: Optional[DeviceMesh], *args, **kwargs):
...
-                                f"arg {arg} in local_map has a mismatched placements: "
-                                f"arg placements is {arg.placements} but the input "
-                                f"placements is {spec}! "
-                                "If redistribute_inputs is wanted, set "
-                                "redistribute_inputs=True to local_map."
-                            )
+                    # convert to tensor to force grad_placements on the original tensor
+                    arg = arg.to_local(grad_placements=spec)
+                    # convert back to Dtensor to redistribute
+                    arg = maybe_redistribute(arg, device_mesh, placement_orig, True)
```
Comment (wanchaol) [inline torch/distributed/tensor/experimental/_func_map.py]: hmmm why it requires converting back to DTensor and perform redistribute here? IMO `in_grad_placements` is just a hint that passed down to `to_local` as `grad_placements` (see explanation [here](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_api.py#L453)), there should be no need to redistribute the arg here

File: torch/distributed/tensor/experimental/_func_map.py [hunk 7]
```diff
@@ -155,27 +199,32 @@ def wrapped(device_mesh: Optional[DeviceMesh], *args, **kwargs):
...
+
+                if in_placements is not None:
+                    spec = in_placements[idx]
+                    arg = maybe_redistribute(
+                        arg, device_mesh, spec, redistribute_inputs
+                    )
 
                 local_arg = arg.to_local()
```
Comment (wanchaol) [inline torch/distributed/tensor/experimental/_func_map.py]: I think what we should do here is: when the in_grad_placements is not None, record it and then just pass the `in_grad_placements` to this `to_local` call. I don't quite understand why there's two `to_local` calls in your PR

File: torch/distributed/tensor/experimental/_func_map.py [hunk 8]
```diff
@@ -127,6 +135,42 @@ def local_map(
...
     """
 
+    def maybe_redistribute(
+        input_tensor: Union[DTensor, torch.Tensor],
+        device_mesh: DeviceMesh,
+        spec: PlacementType,
+        redistribute_inputs: bool = True,
+    ) -> DTensor:
+        """
+        Try redistrubute a tensor. If the input_tensor is not a DTensor, convert to DTensor first.
```
Comment (wanchaol) [inline torch/distributed/tensor/experimental/_func_map.py]: IIUC if the input of this function is not DTensor, we should not convert it to DTensor and do nothing. It looks to me that you should keep the old code path and just handle the in_grad_placements to pass the grad_placements to the `to_local` call below

File: torch/distributed/tensor/experimental/_func_map.py [hunk 9]
```diff
@@ -66,6 +67,13 @@ def local_map(
             will be skipped and the argument will be directly passed to ``func``.
             If ``in_placements`` is ``None``, no placements examination will be performed.
             Default: None
+        in_grad_placements (Tuple[`PlacementType`, ...], optional):
+            the required placements of the :class:`DTensor` s gradient during back backpropagation.
```
Comment (wanchaol) [inline torch/distributed/tensor/experimental/_func_map.py]: The comment here is pretty confusing, I think It should be sth like: 

the placements hint of the :class:`DTensor` s gradient corresponds to the flattened input DTensor. This argument is the hint that user can give to :meth:`to_local` in case the gradient layout of the local tensor input does not match its DTensor input layout.
If not specified, we will assume the gradient layout of the local tensor input remains the same as the original DTensor input and use that for gradient computation.

File: test/distributed/tensor/experimental/test_local_map.py [hunk 10]
```diff
@@ -318,6 +324,66 @@ def test_local_map_redistribute(self):
...
+                with torch.no_grad():
+                    X_dt_grad_corrected = X_dt.grad.from_local(
+                        X_dt.grad.to_local(), device_mesh, placements=X_dt.placements
+                    )
+                    self.assertEqual(X.grad, X_dt_grad_corrected.full_tensor())
+                    W_dt_grad_corrected = W_dt.grad.from_local(
+                        W_dt.grad.to_local(), device_mesh, placements=W_dt.placements
+                    )
+                    self.assertEqual(W.grad, W_dt_grad_corrected.full_tensor())
+
```
Comment (wanchaol) [inline test/distributed/tensor/experimental/test_local_map.py]: I wonder what is the testing intention of this part of code?

File: test/distributed/tensor/experimental/test_local_map.py [hunk 11]
```diff
@@ -318,6 +324,66 @@ def test_local_map_redistribute(self):
...
+            loss.backward()
+            self.assertIsInstance(W_dt.grad, DTensor)
+
+            self.assertEqual(X_dt.grad.placements, in_grad_placements[0])
+            self.assertEqual(W_dt.grad.placements, in_grad_placements[1])
+            if in_grad_placements == (X_dt.placements, Y_dt.placements):
+                self.assertEqual(X_dt.grad.full_tensor(), X.grad)
+                self.assertEqual(W_dt.grad.full_tensor(), W.grad)
+            else:
+                # intentionally set the mismatched `in_grad_placements` with the
```
Comment (wanchaol) [inline test/distributed/tensor/experimental/test_local_map.py]: I think this test case should test when the correct behavior of in_grad_placements. sth similar to the DP case that @tianyu-l mentioned. where:
* two input arg, one is Shard(0), one is replicate
* the output should be Shard(0)
* the in_grad_placements should be Partial

**Review by wanchaol — CHANGES_REQUESTED** [review]
Thanks for adding this feature! The implementation looks not quite right, please see inlined comments.

File: torch/distributed/tensor/experimental/_func_map.py (hunk 7 above)
Comment (zpcore) [reply torch/distributed/tensor/experimental/_func_map.py]: Thanks for the feedback. I have removed the extra `to_local` call with just `local_arg = arg.to_local(grad_placements=spec)`.

File: torch/distributed/tensor/experimental/_func_map.py (hunk 8 above)
Comment (zpcore) [reply torch/distributed/tensor/experimental/_func_map.py]: Extra `to_local` has been removed now.

File: test/distributed/tensor/experimental/test_local_map.py (hunk 10 above)
Comment (zpcore) [reply test/distributed/tensor/experimental/test_local_map.py]: I have updated the test to do something similar to DDP with sharding on input X. Now the `in_grad_placements` makes much sense now. PTAL, thanks!

File: test/distributed/tensor/experimental/test_local_map.py (hunk 11 above)
Comment (zpcore) [reply test/distributed/tensor/experimental/test_local_map.py]: Thanks for the reference! I now use the matmul `X@W --> Y` in the test, with a sharding similar to DP:
- `X_dt` is Shard(0), `W_dt` is replicated
-  `Y` 's placement is Shard(0)
- In the backward, `X_dt` in_grad_placement is Shard(0) and `W_dt` in grad_placement should be Partial().

`X_dt.grad.full_tensor()` and `W_dt.grad.full_tensor()` are expected to be correct under this `in_grad_placement` spec.

**Review by wanchaol — APPROVED** [review]
looks great, thanks for addressing the comments!
